sklearn/cluster/_bisect_k_means.py

```diff
@@ -0,0 +1,727 @@
+
+        # Only assign to created centroid when n_clusters == 1
+        if self.n_clusters == 1:
+            x_squared_norms = row_norms(X, squared=True)
+
+            clusters = self._init_centroids(
+                X, x_squared_norms, init, random_state, n_centroids=1
+            )
+            warnings.warn(
+                "Bisection won't be performed - needs at least two clusters to run."
+            )
+
+            self.labels_ = np.zeros(X.shape[0], dtype=np.intc)
+
+        else:
+            # Run proper bisection to gather
+            # self.cluster_centers_ and self.labels_
+            clusters, self.labels_ = self._run_bisect_kmeans(
+                X, sample_weight, random_state
+            )
+
+        # Restore original data
+        if not sp.issparse(X):
+            X += self._X_mean
+            clusters += self._X_mean
+
+        self.cluster_centers_ = clusters
+
+        self.inertia_ = _inertia(
+            X, sample_weight, self.cluster_centers_, self.labels_, self._n_threads
+        )
+
+        return self
+
+    def _run_bisect_kmeans(self, X, sample_weight, random_state):
+        """Performs Bisecting K-Means, which splits always cluster depending
+        on 'bisect_strategy' attribute:
+
+        - "biggest sse": Picks cluster with biggest SSE (Sum of Squared Errors)
+         from all calculated.
+
+        - "largest_cluster": Picks always cluster with largest number of
+         points assigned from all calculated. That method will perform faster
+          than picking by SSE methods, while producing similar results.
+
+        .. note:: All of passed parameters must be pre-calculated
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            Training instances to cluster.
+
+        sample_weight : array-like of shape (n_samples,)
+            The weights for each observation in X.
+
+        random_state : int, RandomState instance
+            Determines random number generation for centroid initialization.
+
+        Returns
+        ----------
+        cluster_centers_ : ndarray of shape (n_clusters, n_features)
+            Coordinates of cluster centers.
+
+        labels_ : ndarray of shape (n_samples,)
+            Labels of each point.
+        """
+        strategy = self.bisect_strategy
+
+        # Dictionary to imitate tree view of created centers.
+        # Used to keep hierarchical ordering.
+        tree_dict = {-1: {"children": None, "center": None}}
+
+        # Leaves Dictionary used for clustering.
+        # Stores information which data points are assigned to given cluster
+        # along with error or amount of assigned points, depending on
+        # specified 'bisect_strategy'
+        leaves_dict = {
+            -1: {"samples": np.ones(X.shape[0], dtype=bool), "error_or_size": None}
+        }
+
+        # ID of biggest center stored in centers_dict
+        biggest_id = -1
```

Comment:
For consistency with `child_id`, I would rename it `parent_id`.

Reply:
I was more thinking here to mark that its currently the biggest cluster according to chosen criteria. That should be Ok too